compiler-core/src/analyse/name.rs

```diff
@@ -0,0 +1,139 @@
+use ecow::{eco_format, EcoString};
+use regex::Regex;
+
+use crate::ast::{ArgNames, SrcSpan};
+
+use super::{BadNameKind, Error};
+use heck::{ToSnakeCase, ToUpperCamelCase};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NameCorrection {
+    pub location: SrcSpan,
+    pub correction: EcoString,
+}
+
+pub fn check_valid_name(
+    location: SrcSpan,
+    name: &EcoString,
+    kind: BadNameKind,
+) -> Option<(Error, NameCorrection)> {
+    // Some of the internally generated variables (such as `_capture` and `_use0`)
+    // start with underscores, so we allow underscores here.
+    let valid_name_regex = Regex::new("^_?[a-z][a-z0-9_]*$").expect("Regex is correct");
```

Comment:
This should be a static value rather than recomputing the regex on every name. See other uses of regex in the compiler for how we do this